source/extensions/filters/http/grpc_json_transcoder/json_transcoder_filter.cc

```diff
@@ -107,114 +118,187 @@ class TranscoderImpl : public Transcoder {
+    }
+    Envoy::Grpc::AsyncStream<ServerReflectionRequest> async_stream =
+        async_client->start(*server_reflection_method, *this, options);
+
+    ServerReflectionRequest request;
+    request.set_file_containing_symbol(service);
+    async_stream.sendMessage(request, true);
   }
+}
 
-  FileDescriptorSet descriptor_set;
+void AsyncReflectionFetcher::onReceiveMessage(std::unique_ptr<ServerReflectionResponse>&& message) {
+  std::lock_guard<std::mutex> guard_(receive_message_mutex_);
+  const std::string& service_requested = message->original_request().file_containing_symbol();
 
-  switch (proto_config.descriptor_set_case()) {
-  case envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder::
-      DescriptorSetCase::kProtoDescriptor:
-    if (!descriptor_set.ParseFromString(
-            api.fileSystem().fileReadToEnd(proto_config.proto_descriptor()))) {
-      throw EnvoyException("transcoding_filter: Unable to parse proto descriptor");
-    }
-    break;
-  case envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder::
-      DescriptorSetCase::kProtoDescriptorBin:
-    if (!descriptor_set.ParseFromString(proto_config.proto_descriptor_bin())) {
-      throw EnvoyException("transcoding_filter: Unable to parse proto descriptor");
+  // This was not a requested service, so assume this is a duplicate response and ignore it.
+  if (services_.find(service_requested) == services_.end()) {
+    return;
+  }
+
+  for (const std::string& file_descriptor_bytes:
+      message->file_descriptor_response().file_descriptor_proto()) {
+    if (!file_descriptor_set_.add_file()->ParseFromString(file_descriptor_bytes)) {
+      throw EnvoyException(
+          "transcoding_filter: Unable to parse proto descriptor returned from reflection");
     }
-    break;
-  case envoy::extensions::filters::http::grpc_json_transcoder::v3::GrpcJsonTranscoder::
-      DescriptorSetCase::DESCRIPTOR_SET_NOT_SET:
-    throw EnvoyException("transcoding_filter: descriptor not set");
   }
 
-  for (const auto& file : descriptor_set.file()) {
-    addFileDescriptor(file);
+  // Record that we have already received the response for this service.
+  services_.erase(service_requested);
+
+  // We have received all the protos we have asked for, so inject them into the
+  // proto config and signal to InitManager that loading is complete.
+  if (services_.empty()) {
+    descriptor_pool_builder_->loadFileDescriptorProtos(file_descriptor_set_);
+    init_target_->ready();
   }
+}
 
-  convert_grpc_status_ = proto_config.convert_grpc_status();
-  if (convert_grpc_status_) {
-    addBuiltinSymbolDescriptor("google.protobuf.Any");
-    addBuiltinSymbolDescriptor("google.rpc.Status");
+void AsyncReflectionFetcher::onReceiveTrailingMetadata(Http::ResponseTrailerMapPtr&&) {}
+void AsyncReflectionFetcher::onRemoteClose(Envoy::Grpc::Status::GrpcStatus status, const std::string& message) {
+  if (status != Envoy::Grpc::Status::Ok) {
```

Comment:
Failure handling is tricky,  1) we could not thow exception.   2) we need to know all our requests have been processed. success or failure.   My suggestion is to create separate callback instance (used in start call in line 152)  for each remote call.  Only remember these callback,  when it is called, remove it from the list.  In this way, we will know if all of them have been processed.